### PR TITLE
Feature: custom annotation in Web VEP (e112)

### PR DESCRIPTION
**Pull request by nakib103**
[ENSVAR-2739](https://www.ebi.ac.uk/panda/jira/browse/ENSVAR-2739)

Two bugfix in web VEP -
1. Currently there are section added in the input form which are empty and is confusing/does not look good -
image

Added a fix to only add section to relevant species.

2. There is a glitch present in `Variant synonyms` option. In **Find co-located known variants:** option, if you select 'No' and then 'yes', `Variant synonyms` option will appear regardless of species, e.g. in goat - 

image

Fixed this bug, by basically wrapping the option in a `div` element and adding the filter to `_stt_Homo_sapiens` and `__stt_Sus_scrofa` to this parent element.

### Test

Frequency:
sheep nextgen: http://wp-np2-11.ebi.ac.uk:7070/Ovis_aries_rambouillet/Tools/VEP/Ticket?tl=NZtMMA0oIiFkeYSW
sheep texel nextgen: http://wp-np2-11.ebi.ac.uk:7070/Ovis_aries/Tools/VEP/Results?tl=9OHpITBFw9iFN9yG-78
goat nextgen: http://wp-np2-11.ebi.ac.uk:7070/Ovis_aries_rambouillet/Tools/VEP/Ticket?tl=9XRCS6hNaziIVKSI
dog boxer PRJEB24066: http://wp-np2-11.ebi.ac.uk:7070/Ovis_aries_rambouillet/Tools/VEP/Ticket?tl=YBL18ITepCEOhARn
dog EVA PRJEB24066: http://wp-np2-11.ebi.ac.uk:7070/Ovis_aries_rambouillet/Tools/VEP/Ticket?tl=0fA066F8SMpjstKJ
chicken PRJEB44919: http://wp-np2-11.ebi.ac.uk:7070/Ovis_aries_rambouillet/Tools/VEP/Ticket?tl=pKuUZa8ps7Vd5uMY

Regulatory:
turbot: http://wp-np2-11.ebi.ac.uk:7070/Scophthalmus_maximus/Tools/VEP/Results?tl=RVFGuPYUBWKUkga5-75
european seabass: http://wp-np2-11.ebi.ac.uk:7070/Dicentrarchus_labrax/Tools/VEP/Results?tl=fC8NJcUfZONXWdF1-76

Generic (Proof-of-concept)
ClinVar: http://wp-np2-11.ebi.ac.uk:7070/Homo_sapiens/Tools/VEP/Results?tl=NU2bnPy6c7bJNShN-62

### CODE REVIEW

**Comment by imransl** [comment]
@nuno-agostinho Can you please review/look into this PR?

File: tools/modules/EnsEMBL/Web/Object/VEP.pm [hunk 1]
```diff
@@ -502,7 +502,7 @@ sub species_list {
         'variation'   => $db_config->{'DATABASE_VARIATION'} // undef,
         'refseq'      => $db_config->{'DATABASE_OTHERFEATURES'} && $sd->get_config($_, 'VEP_REFSEQ') // undef,
         'assembly'    => $sd->get_config($_, 'ASSEMBLY_NAME') // undef,
-        'regulatory'  => $sd->get_config($_, 'REGULATORY_BUILD') // undef,
+        'regulatory'  => $db_config->{'DATABASE_FUNCGEN'}{'tables'}{'regulatory_build'}{'analyses'}{'Regulatory_Build'}->{'count'} // undef,
```
Comment (nuno-agostinho) [inline tools/modules/EnsEMBL/Web/Object/VEP.pm:505]: Why the need to change this?

File: tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm [hunk 2]
```diff
@@ -1034,4 +1099,53 @@ sub _add_plugins {
...
+  my $species       = $self->object->species_list;
+
+  foreach my $custom(@{$self->_get_customs_by_section($section_name)}) {
+    my $custom_id = $custom->{id};
+
+    my $field_class;
+    if (!$custom->{species} || $custom->{species} eq '*') {
+      $field_class = [];
+    } else {
+      # Avoid showing specific plugins if supported species are not available
```
Comment (nuno-agostinho) [inline tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm:1133]:
```suggestion
      # Avoid showing if supported species are not available
```

File: tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm [hunk 3]
```diff
@@ -514,6 +518,35 @@ sub _build_variants_frequency_data {
       })]
     }) if (first { $_->{'value'} eq 'Homo_sapiens' } @$species);
 
+    my @custom_frequencies = @{$self->_get_customs_by_section($current_section)};
+    my $frequncy_species_data = {};
```
Comment (nuno-agostinho) [inline tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm:522]: Misspelled variable, could you please fix this variable in this file?
```suggestion
    my $frequency_species_data = {};
```

**Review by nuno-agostinho — CHANGES_REQUESTED** [review]
Hi @nakib103, thanks for your changes. Seems to be working as intended, although I have a few suggestions/questions.

My major concern is related with the checkboxes labels in web VEP form and tooltips for the columns in the results table, specially compared to the human frequency options already available.

## Web VEP form

Similar to how human presents data, the labels should be self-explanatory.

Screenshot 2024-02-13 at 13 15 21

Also, when there are allele frequencies for multiple populations, the tooltip details all the populations with both their acronym and a small description in parenthesis:

Screenshot 2024-02-13 at 13 24 19

I think this should be improved on a case-by-case basis:
- For EVA studies, I would change the label to `EVA study PRJEB44919 allele frequencies` (the tooltip seems fine)
- For goat, `NextGen Project allele frequencies` (the tooltip must also describe the [goat populations](https://www.ensembl.org/info/genome/variation/species/populations.html#Capra_hircus))
- For sheep, we could display:
    - `NextGen Project (Iranian) allele frequencies` (tooltip: `Frequency data for co-located variant from NextGen project (NextGen:IROA)`)
    - `NextGen Project (Moroccan) allele frequencies` (tooltip: `Frequency data for co-located variant from NextGen project (NextGen:MOOA)`)

## Table results

For human, allele frequencies column names show a tooltip describing the population in details. This is specially important for the multiple populations in goat.

Screenshot 2024-02-13 at 13 23 01

Also, the first column seems to be the ID of the variant in the VCF, which is not always useful (sometimes it's just coordinates).  Anyway, the name of the column should contain "ID" in the end to be easier to understand what it is.

File: tools/modules/EnsEMBL/Web/Object/VEP.pm (hunk 1 above)
Comment (nakib103) [reply tools/modules/EnsEMBL/Web/Object/VEP.pm]: Hi @nuno-agostinho , this is for automated detection of species that have regulatory build. Currently we rely on INI file but with this change we do not have to update ini file manually, web VEP will automatically pick up species that we can run `--regulatory` on.

I have removed these changes for now - as after discussion with Garth, they would review the config and make update if needed (basically we would not rely on `analyses` table in future). We can automate this part when we get confirmation. The [PR](https://github.com/Ensembl/public-plugins/pull/766) is targeted for e113 now.

File: tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm (hunk 2 above)
Comment (nakib103) [reply tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm]: updated! thanks Nuno

File: tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm (hunk 3 above)
Comment (nakib103) [reply tools/modules/EnsEMBL/Web/Component/Tools/VEP/InputForm.pm]: updated, thanks Nuno!

**Comment by nakib103** [comment]
Hi @nuno-agostinho ,

About helptip and description -
- currently the labels come directly from `short_name`. I have added a prefix `allele frequencies`. It is not exactly same as you mentioned but more clear. Any more descriptive is not possible from current conifg (I can add another key/value pair but that seems a bit redundant)
- instead, the description would make explain it more. The description value can be as large as need be. It explains what `IROA` or `MOOA` population means, or goat populations included.
- also added helptip in the output. This is achieved by adding `helptips` in the config.

**Review by nuno-agostinho — APPROVED** [review]
Thanks for the changes, @nakib103! Everything is looking good to me!

**Pull request merged**